plugin/proxy/proxy.go

```diff
@@ -116,19 +115,6 @@ func (p Proxy) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (
 				return 0, taperr
 			}
 
-			// A "ANY isc.org" query is being dropped by ISC's nameserver, we see this as a i/o timeout, but
-			// would then mark our upstream is being broken. We should not do this if we consider the error temporary.
-			// Of course it could really be that our upstream is broken
-			if oe, ok := backendErr.(*net.OpError); ok {
-				// Note this keeps looping and trying until tryDuration is hit, at which point our client
-				// might be long gone...
-				if oe.Timeout() {
-					// Our upstream's upstream is problably messing up, continue with next selected
-					// host - which my be the *same* one as we don't set any uh.Fails.
-					continue
-				}
-			}
-
 			timeout := host.FailTimeout
 			if timeout == 0 {
```

Comment:
For this change, I am hesitant.
I do not know why this code was introduced.
We need validation of @miekg  here.

Reply:
After talk and review, if we remove these lines, the impact will be that this upstream will be "failed" during the "fail timeout" period.

- That will allow to choose the next Upstream in case of "sequential" policy
- That will not hit other use case as anyway another Upstream will be choose.

So maybe that code was introduced before the mechanism of Down/failtimeout

Reply:
You can't rip this out; and this piece of code shows that we should move to *forward* as the default proxy impl. As we attribute upstream dead nameservers to *our* local proxy. The forward plugin fires off a healthcheck in this case to re-asses (even there is can be a bit smarter).

Reply:
@miekg : what is the purpose of this code ? 
I mean : we want to avoid the healthcheck if there is a timeout ? 
if YES, is that still valid after the fix #1441 ?

What I mean is if we get rid of this code:
-  a timeout will be considered as a "fail" for 2 sec, 
until either there is a healtcheck. 
- if healhcheck is not defined, or healtcheck is ok, this upstream will be available again for next Select().

Reply:
Maybe I misunderstand the point of the logic around this timeout. Let me know.

Reply:
Is there a test you can add that shows the issue which will be fixed by this PR?

Reply:
Good point. Let's make a UT to validate before and after. Thanks